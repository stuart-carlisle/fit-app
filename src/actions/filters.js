
// SET_DATE_FILTER
export const setDateFilter = (date) => ({
  type: 'SET_DATE_FILTER',
  date
})

// SET_TEXT_FILTER
export const setTextFilter = (text = '') => ({
  type: 'SET_TEXT_FILTER',
  text
});

// SORT_BY_DATE
export const sortByDate = () => ({
  type: 'SORT_BY_DATE'
});

// SORT_BY_ALPHABETICAL
export const sortByAlphabetical = () => ({
  type: 'SORT_BY_ALPHABETICAL'
});

// SET_REFERENCE_DATABASE_TO_PUBLIC
export const setReferenceDatabaseToPublic = () => ({
  type: 'SET_REFERENCE_DATABASE_TO_PUBLIC'
})

// SET_REFERENCE_DATABASE_TO_PRIVATE
export const setReferenceDatabaseToPrivate = () => ({
  type: 'SET_REFERENCE_DATABASE_TO_PRIVATE'
})

// SET_CHART_TYPE
export const setChartType = (chartType) => ({
  type: 'SET_CHART_TYPE',
  chartType
})

// SET_MEAL_FILTER
export const setMealFilter = (meal) => ({
  type: 'SET_MEAL_FILTER',
  meal
})

