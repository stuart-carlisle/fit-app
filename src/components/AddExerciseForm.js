import React, { useState , useEffect} from 'react'
import moment from 'moment'
import { connect } from 'react-redux'
import numeral from 'numeral'

export const AddExerciseForm = ({
  filters = {},
  onSubmit, 
  exercise = {
    description:'',
    type:'',
    energy:'',
    time:''
  }
}) => {
    const defaultEnergy = numeral(exercise.energy).value()
    const defaultTime = numeral(exercise.time).value()
    const date = filters.date
    const referenceDatabase = filters.referenceDatabase
    const [time, setTime] = useState(defaultTime)
    const [error, setError] = useState('')
    const [energy, setEnergy] = useState(defaultEnergy)
    const diaryDate = date ? date.valueOf() : moment().startOf('day').valueOf()
    const lastUsed = moment().valueOf()

    useEffect(()=>{
        
      let refreshed = JSON.parse(localStorage.getItem('refreshed'))
      if(refreshed){
        window.history.back()
        window.scrollTo(0,0)
        localStorage.removeItem('refreshed')
      }else{
        localStorage.setItem('refreshed',JSON.stringify(true))
      }
    },[])

    const onFormSubmit = (e) => {
        e.preventDefault();
        const exerciseData = { 
          description: exercise.description,
          type: exercise.type
        }
        if (numeral(time).value()<=0) {
          setError('Please provide a time greater than 0')
          window.scrollTo(0,0)
        } else {
          setError('')
          onSubmit(
            referenceDatabase,
            diaryDate,
            lastUsed,
            exerciseData,
            time.toString(),
            energy.toString()
          )
          localStorage.removeItem('refreshed')
        }
      }
    
    const onTimeChange = (e)=>{
        const time = e.target.value
          if (!time || time.match(/^\d{1,}(\.\d{0,1})?$/)) {
          setTime(time)
          const timeValue = numeral(time).value()
          setEnergy(Math.round(defaultEnergy * (timeValue/defaultTime)))
        }
    }

    const onEnergyChange = ()=>{
      
    }

    return (
        <div>
        <div className="content-container__background">
        <div className="content-container__form__update">
            <form className= "form" onSubmit={onFormSubmit}>
                {error && <p id="error-message" className="form__error">{error}</p>}
                  <h1 className="page-header__title-update"> {exercise.description} </h1>
                    <span className="text-input__span-container-form">
                      <div className="content-container__update-titles">
                        <h3 className="form__heading__update">Exercise Time (mins):</h3>
                        <h3 className="form__heading__update">Energy (calories): </h3>
                      </div>
                      <div className="content-container__update">
                        <input  //Exercise Duration
                          className="text-input__update-form"
                          type="text"
                          value={time}
                          onChange={onTimeChange}
                          placeholder="Enter a Duration"
                        />  
                        <input  //Energy Used
                          className="text-input__update-form"
                          type="text"
                          value={energy?energy:''}
                          onChange={onEnergyChange}
                          placeholder="Energy Used"
                        /> 
                      </div>
                    </span>
                <button className="button button--form">Add Exercise</button>
            </form>
          </div>
        </div>
      </div>  
    )
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}


export default connect(mapStateToProps)(AddExerciseForm);