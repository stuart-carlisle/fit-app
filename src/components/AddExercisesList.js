import React from 'react'
import { connect } from 'react-redux'
import AddExercisesListItem from './AddExercisesListItem'
import selectExercises from '../selectors/exercises'

export const AddExercisesList = ({exercises}) => {
  let exercisesTopResults
  
  if(exercises.length>10){
    exercisesTopResults = exercises.slice(0,10)
  }else{
    exercisesTopResults = exercises
  }

  return(<div className="content-container">
   <div className="list-header">
    <div className="">Exercise</div>
    <div className="show-for-tablet">Calories Burned</div>
    <div className="show-for-mobile">Cals Burned</div>
   </div>
    <div className="list-body">
      {
        exercisesTopResults.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Exercises In History</span>
          </div>
        ) : (
         exercisesTopResults.map((exercise) => {
            return <AddExercisesListItem key={exercise.id} {...exercise} />
          })
        )
      }
    </div>
  </div>  
  )
}

const mapStateToProps = (state) => {
  return {
    exercises: state.filters.referenceDatabase === 'public' ? state.exercisesPublic : selectExercises(state.exercisesPrivate,  state.filters)
  }
}

export default connect(mapStateToProps)(AddExercisesList);
