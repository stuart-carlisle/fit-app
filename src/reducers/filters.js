import moment from 'moment'

// Filters Reducer

const filtersReducerDefaultState = {
  date:moment().startOf('day'),
  text: '',
  sortBy: 'date',
  referenceDatabase: 'private',
  chartType: {
    chart: 'calorie-tracker', 
    option: 'week'
  },
  meal: 'breakfast'
}

export default (state = filtersReducerDefaultState, action) => {
  switch (action.type) {
    case 'SET_DATE_FILTER':
      return{
        ...state,
        date: action.date
      }
    case 'SET_TEXT_FILTER':
      return {
        ...state,
        text: action.text
      };
    case 'SORT_BY_ALPHABETICAL':
      return {
        ...state,
        sortBy: 'alphabetical'
      };
    case 'SORT_BY_DATE':
      return {
        ...state,
        sortBy: 'date'
      };
    case 'SET_REFERENCE_DATABASE_TO_PUBLIC':
      return {
        ...state,
        referenceDatabase: 'public'
      }
    case 'SET_REFERENCE_DATABASE_TO_PRIVATE':
      return {
        ...state,
        referenceDatabase: 'private'
      }
    case 'SET_CHART_TYPE':
      return {
        ...state,
        chartType: action.chartType
      };
    case 'SET_MEAL_FILTER':
      return {
        ...state,
        meal: action.meal
      }
    default:
      return state;
  }
};
