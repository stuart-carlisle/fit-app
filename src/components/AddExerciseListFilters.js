import React, { useState } from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAlphabetical, setReferenceDatabaseToPublic } from '../actions/filters';
import { setExercisesPublic } from '../actions/exercises'
import algoliasearch from 'algoliasearch/lite'

export const AddExerciseListFilters = ({
    filters,
    setTextFilter, 
    setReferenceDatabasePublic,
    sortByDate, 
    sortByAlphabetical,
    setExercisesPublic 
  }) => {
 
  const searchClient = algoliasearch (
      process.env.ALGOLIA_APP_ID,
      process.env.ALGOLIA_API_KEY
  )

  const index = searchClient.initIndex('exercises');

  const [text, setText] = useState('')
    
  const onTextChange = (e) => {
    setTextFilter(e.target.value)
    setText(e.target.value)
  }

  const onSortChange = (e) => {
    if (e.target.value === 'date') {
      sortByDate()
    } else if (e.target.value === 'alphabetical') {
      sortByAlphabetical()
    }
  }

  const onSubmit = (e) => {
    e.preventDefault()
    index
      .search({
        query: text
      })
      .then((responses) => {
        const array = responses.hits
        const newExercises = array.map((exercise)=>{
          return{
            description:exercise.description,
            type:exercise.type,
            energy:exercise.energy,
            time:exercise.time,
            id:exercise.objectID
          }
        })
        setExercisesPublic(newExercises)
      }).then(()=>{
        setReferenceDatabasePublic()
      })
}

    return (
      
      <div className="header__content--filters"> 
      <form className="filters-text-input__span-container" onSubmit={onSubmit}>
          <div className="input-group__item-page-header">
            <input
              type="text"
              className="filters-text-input"
              placeholder="SEARCH EXERCISES"
              value={text}
              onChange={onTextChange}
            />
          </div>
          <div className="input-group__item-page-header">
            <select
              className="minimal"
              value={filters.sortBy}
              onChange={onSortChange}
            >
              <option value="date">DATE</option>
              <option value="alphabetical">ALPHABETICAL</option>
            </select>
          </div>
            <button className="button filters-button--form-page-header">Search Online</button>
            </form>   
        </div>
    )
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAlphabetical: () => dispatch(sortByAlphabetical()),
  setReferenceDatabasePublic: () => dispatch(setReferenceDatabaseToPublic()),
  setExercisesPublic: (exercises) => dispatch(setExercisesPublic(exercises))
})

export default connect(mapStateToProps,mapDispatchToProps)(AddExerciseListFilters);

