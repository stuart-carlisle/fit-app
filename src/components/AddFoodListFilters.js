import React, { useState } from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAlphabetical, setReferenceDatabaseToPublic } from '../actions/filters'
import { setFoodsPublic } from '../actions/foods'
import algoliasearch from 'algoliasearch/lite'

export const AddFoodListFilters = ({
    filters,
    setTextFilter, 
    setReferenceDatabasePublic, 
    sortByDate, 
    sortByAlphabetical,
    setFoodsPublic
  }) => {
  
  const searchClient = algoliasearch (
      process.env.ALGOLIA_APP_ID,
      process.env.ALGOLIA_API_KEY
  )

  const index = searchClient.initIndex('foods');
  
  const [text,setText]=useState('')

  const onTextChange = (e) => {
      setTextFilter(e.target.value)
      setText(e.target.value)
  }

  const onSortChange = (e) => {
    if (e.target.value === 'date') {
      sortByDate()
    } else if (e.target.value === 'alphabetical') {
      sortByAlphabetical()
    }
  }

  const onSubmit = (e) => {
      e.preventDefault()
      index
        .search({
          query: text
        })
        .then((responses) => {
          const array = responses.hits
          const newFoods = array.map((food)=>{
            return{
              description:food.description,
              nutrition:food.nutrition,
              serving:food.serving,
              id:food.objectID
            }
          })
          setFoodsPublic(newFoods)
        }).then(()=>{
          setReferenceDatabasePublic()
        })
  }

    return (
      
      <div className="header__content--filters"> 
      <form className="filters-text-input__span-container" onSubmit={onSubmit}>
          <div className="input-group__item-page-header">
            <input
              type="text"
              className="filters-text-input"
              placeholder="SEARCH FOODS"
              value={filters.text}
              onChange={onTextChange}
            />
          </div>
          <div className="input-group__item-page-header">
            <select
              className="minimal"
              value={filters.sortBy}
              onChange={onSortChange}
            >
              <option value="date">DATE</option>
              <option value="alphabetical">ALPHABETICAL</option>
            </select>
          </div>
            <button className="button filters-button--form-page-header">SEARCH ONLINE</button>
            </form>   
        </div>
   
    )
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAlphabetical: () => dispatch(sortByAlphabetical()),
  setReferenceDatabasePublic: () => dispatch(setReferenceDatabaseToPublic()),
  setFoodsPublic: (foods) => dispatch(setFoodsPublic(foods))
})

export default connect(mapStateToProps,mapDispatchToProps)(AddFoodListFilters);
