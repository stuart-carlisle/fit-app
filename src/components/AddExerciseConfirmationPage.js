import React from 'react'
import { connect } from 'react-redux'
import AddExerciseForm from './AddExerciseForm'
import { Link } from 'react-router-dom'
import { startAddExerciseDiary, startAddExercisePrivate, startUpdateExercisePrivate } from '../actions/exercises'

export const AddExerciseConfirmationPage = (props) => {
  
  const onSubmit = (referenceDatabase, diaryDate, lastUsed, exerciseDescAndType, time, energy) => {
    if (referenceDatabase==='public'){
      props.startAddExercisePrivate({...exerciseDescAndType, time, energy, lastUsed})
    } else if (referenceDatabase==='private'){
      props.startUpdateExercisePrivate(props.exercise.id,{...exerciseDescAndType, time, energy, lastUsed})
    }
    props.startAddExerciseDiary({...exerciseDescAndType, time, diaryDate, energy})
    props.history.push('/diary');
  }
  

  return (
     <div>
     <Link className="button__back-button" to={`/add-exercise`}><img className="image__header" src="/images/back-arrow.svg"/></Link>
        <div className="page-header__filter page-header__filter--with-back">
          <div className="content-container">
            <h1 className="page-header__title">ADD EXERCISE</h1>
          </div>
        </div>
        <div className="show-for-mobile page-header__margin"></div>
        <div className="content-container">
        <AddExerciseForm id="add-exercise-form" exercise={props.exercise} onSubmit={onSubmit} />
        </div>
     </div>
 )
}

const mapStateToProps = (state,props) => {
  if(state.filters.referenceDatabase==='public'){
    return {
      exercise: state.exercisesPublic.find((exercise) => exercise.id === props.match.params.id)
    }
  }else if(state.filters.referenceDatabase==='private'){
    return {
      exercise: state.exercisesPrivate.find((exercise) => exercise.id === props.match.params.id)
    }
  }
}

const mapDispatchToProps = (dispatch) =>({
    startAddExerciseDiary: (exercise) => dispatch(startAddExerciseDiary(exercise)),
    startUpdateExercisePrivate: (id, exercise) => dispatch(startUpdateExercisePrivate(id, exercise)),
    startAddExercisePrivate: (exercise) => dispatch(startAddExercisePrivate(exercise))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddExerciseConfirmationPage);




