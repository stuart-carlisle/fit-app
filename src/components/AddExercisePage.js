import React, { useEffect } from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import AddExerciseListFilters from './AddExerciseListFilters'
import AddExercisesList from './AddExercisesList'
import { setTextFilter, sortByDate, setReferenceDatabaseToPrivate } from '../actions/filters'

export const AddExercisePage = ({ setTextFilter, sortByDate, setReferenceDatabaseToPrivate}) => {
  useEffect(()=>{
    localStorage.removeItem('refreshed')
      setTextFilter('')
      sortByDate()
      setReferenceDatabaseToPrivate()
  },[])
    return(
      <div>
      <Link className="button__back-button" to={`/diary`}><img className="image__header" src="/images/back-arrow.svg"/></Link>
    <div className="page-header__filter page-header__filter--with-back">
      <div className="content-container">
        <AddExerciseListFilters />
      </div>  
    </div>
    <AddExercisesList />
    <div className="content-container">
    <Link className="button button--link" to={`/create-exercise`}>+ CREATE CUSTOM EXERCISE</Link>
    </div>
    </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  setReferenceDatabaseToPrivate: () => dispatch(setReferenceDatabaseToPrivate())
})

export default connect (undefined, mapDispatchToProps)(AddExercisePage)