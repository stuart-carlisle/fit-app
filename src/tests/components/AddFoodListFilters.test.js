import React from 'react'
import { shallow } from 'enzyme'
import { AddFoodListFilters } from '../../components/AddFoodListFilters'
import filters, { altFilters } from '../fixtures/filters'

let setTextFilter, setReferenceDatabasePublic, sortByDate, sortByAlphabetical, setFoodsPublic, wrapper

beforeEach(()=>{
    setTextFilter=jest.fn()
    setReferenceDatabasePublic=jest.fn()
    sortByDate=jest.fn()
    sortByAlphabetical=jest.fn()
    setFoodsPublic=jest.fn() 
    wrapper = shallow(
        <AddFoodListFilters
            filters={filters}
            setTextFilter={setTextFilter}
            setReferenceDatabasePublic={setReferenceDatabasePublic}
            sortByDate={sortByDate}
            sortByAlphabetical={sortByAlphabetical}
            setFoodsPublic={setFoodsPublic}
        />
    )
})

test('Should render add food list filters correctly',()=>{
    expect(wrapper).toMatchSnapshot()
})

test('Should render add food list alt filters correctly',()=>{
    wrapper.setProps({
        filters:altFilters
    })
    expect(wrapper).toMatchSnapshot()
})

test('Should handle text change', ()=>{
    const value = 'run'
    wrapper.find('input').simulate('change',{
        target:{
            value
        }
    })
    expect(setTextFilter).toHaveBeenLastCalledWith(value)
})

test('Should sort by date', ()=>{
    const value = 'date'
    wrapper.setProps({
        filter: altFilters
    })
    wrapper.find('select').simulate('change',{
        target:{
            value
        }
    })
    expect(sortByDate).toHaveBeenCalled()
})

test('Should sort by alphabetical', ()=>{
    const value = 'alphabetical'
    wrapper.find('select').simulate('change',{
        target:{
            value
        }
    })
    expect(sortByAlphabetical).toHaveBeenCalled()
})

//need to make a dummy algolia search in the future to test the form submission