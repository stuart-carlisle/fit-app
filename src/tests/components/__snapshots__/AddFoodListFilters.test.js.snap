// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render add food list alt filters correctly 1`] = `
<div
  className="header__content--filters"
>
  <form
    className="filters-text-input__span-container"
    onSubmit={[Function]}
  >
    <div
      className="input-group__item-page-header"
    >
      <input
        className="filters-text-input"
        onChange={[Function]}
        placeholder="SEARCH FOODS"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group__item-page-header"
    >
      <select
        className="minimal"
        onChange={[Function]}
        value="alphabetical"
      >
        <option
          value="date"
        >
          DATE
        </option>
        <option
          value="alphabetical"
        >
          ALPHABETICAL
        </option>
      </select>
    </div>
    <button
      className="button filters-button--form-page-header"
    >
      SEARCH ONLINE
    </button>
  </form>
</div>
`;

exports[`Should render add food list filters correctly 1`] = `
<div
  className="header__content--filters"
>
  <form
    className="filters-text-input__span-container"
    onSubmit={[Function]}
  >
    <div
      className="input-group__item-page-header"
    >
      <input
        className="filters-text-input"
        onChange={[Function]}
        placeholder="SEARCH FOODS"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group__item-page-header"
    >
      <select
        className="minimal"
        onChange={[Function]}
        value="date"
      >
        <option
          value="date"
        >
          DATE
        </option>
        <option
          value="alphabetical"
        >
          ALPHABETICAL
        </option>
      </select>
    </div>
    <button
      className="button filters-button--form-page-header"
    >
      SEARCH ONLINE
    </button>
  </form>
</div>
`;
